gtsam/slam/dataset.cpp

```diff
@@ -73,88 +79,167 @@ string findExampleDataFile(const string& name) {
-            + name);
+        "gtsam::createRewrittenFileName could not find a matching file in\n" +
+        name);
   }
 
   fs::path p(name);
-  fs::path newpath = fs::path(p.parent_path().string())
-      / fs::path(p.stem().string() + "-rewritten.txt");
+  fs::path newpath = fs::path(p.parent_path().string()) /
+                     fs::path(p.stem().string() + "-rewritten.txt");
 
   return newpath.string();
 }
 
 /* ************************************************************************* */
-GraphAndValues load2D(pair<string, SharedNoiseModel> dataset, int maxID,
-    bool addNoise, bool smart, NoiseFormat noiseFormat,
-    KernelFunctionType kernelFunctionType) {
-  return load2D(dataset.first, dataset.second, maxID, addNoise, smart,
-      noiseFormat, kernelFunctionType);
+// Parse a file by calling the parse(is, tag) function for every line.
+template <typename T>
+using Parser =
+    std::function<boost::optional<T>(istream &is, const string &tag)>;
+
+template <typename T>
+static void parseLines(const string &filename, Parser<T> parse) {
+  ifstream is(filename.c_str());
+  if (!is)
+    throw invalid_argument("parse: can not find file " + filename);
+  while (!is.eof()) {
+    string tag;
+    is >> tag;
+    parse(is, tag); // ignore return value
+    is.ignore(LINESIZE, '\n');
+  }
+}
+
+/* ************************************************************************* */
+// Parse types T into a size_t-indexed map
+template <typename T>
+map<size_t, T> parseToMap(const string &filename, Parser<pair<size_t, T>> parse,
+                          size_t maxIndex) {
+  map<size_t, T> result;
+  Parser<pair<size_t, T>> emplace = [&](istream &is, const string &tag) {
+    if (auto t = parse(is, tag)) {
+      if (!maxIndex || t->first <= maxIndex)
+        result.emplace(*t);
+    }
+    return boost::none;
+  };
+  parseLines(filename, emplace);
+  return result;
 }
 
 /* ************************************************************************* */
-// Read noise parameters and interpret them according to flags
-static SharedNoiseModel readNoiseModel(ifstream& is, bool smart,
-    NoiseFormat noiseFormat, KernelFunctionType kernelFunctionType) {
-  double v1, v2, v3, v4, v5, v6;
-  is >> v1 >> v2 >> v3 >> v4 >> v5 >> v6;
+// Parse a file and push results on a vector
+template <typename T>
+static vector<T> parseToVector(const string &filename, Parser<T> parse) {
+  vector<T> result;
+  Parser<T> add = [&result, parse](istream &is, const string &tag) {
+    if (auto t = parse(is, tag))
+      result.push_back(*t);
+    return boost::none;
+  };
+  parseLines(filename, add);
+  return result;
+}
+
+/* ************************************************************************* */
+boost::optional<IndexedPose> parseVertexPose(istream &is, const string &tag) {
+  if ((tag == "VERTEX2") || (tag == "VERTEX_SE2") || (tag == "VERTEX")) {
+    size_t id;
+    double x, y, yaw;
+    is >> id >> x >> y >> yaw;
```

Comment:
Would be nice to detect malformed lines:
```
if(!is >> id >> x >> y >> yaw) {
// throw or return boost::none
}
```

Reply:
I did this for the 3 parseXXX routines.